File: Dddml.Wms.JavaServices/src/main/java/org/dddml/wms/domain/InOutDocumentActionCommandHandler.java
package org.dddml.wms.domain;

import org.dddml.wms.specialization.CommandType;
import org.dddml.wms.specialization.PropertyCommand;
import org.dddml.wms.specialization.PropertyCommandHandler;
import org.squirrelframework.foundation.fsm.StateMachineBuilder;
import org.squirrelframework.foundation.fsm.StateMachineBuilderFactory;
import org.squirrelframework.foundation.fsm.impl.AbstractStateMachine;

/**
 * Created by Yang on 2016/7/31.
 */
public class InOutDocumentActionCommandHandler implements PropertyCommandHandler<String, String>
{

    public static class DocumentStatusStateMachine extends AbstractStateMachine<DocumentStatusStateMachine, String, String, Object>
    {

    }

    public static StateMachineBuilder<DocumentStatusStateMachine, String, String, Object> documentStatusStateMachineBuilder;

    static
    {
        documentStatusStateMachineBuilder
                = StateMachineBuilderFactory.create(DocumentStatusStateMachine.class, String.class,String.class, Object.class);
        documentStatusStateMachineBuilder.externalTransition().from(DocumentStatusIds.INITIAL).to(DocumentStatusIds.DRAFTED).on(DocumentActionName.DRAFT);
        documentStatusStateMachineBuilder.externalTransition().from(DocumentStatusIds.DRAFTED).to(DocumentStatusIds.COMPLETED).on(DocumentActionName.COMPLETE);
        documentStatusStateMachineBuilder.externalTransition().from(DocumentStatusIds.DRAFTED).to(DocumentStatusIds.VOIDED).on(DocumentActionName.VOID);
        documentStatusStateMachineBuilder.externalTransition().from(DocumentStatusIds.COMPLETED).to(DocumentStatusIds.CLOSED).on(DocumentActionName.CLOSE);
        documentStatusStateMachineBuilder.externalTransition().from(DocumentStatusIds.COMPLETED).to(DocumentStatusIds.REVERSED).on(DocumentActionName.REVERSE);

    }

    @Override
    public final void execute(PropertyCommand<String, String> command)
    {
        String currentState = command.getStateGetter().get();
        String trigger = command.getContent();

        if (command.getOuterCommandType().equals(CommandType.CREATE)) {
            if (currentState == null || currentState.trim().equals("")) {
                currentState = DocumentStatusIds.INITIAL;
            }
            if (trigger == null) {
                trigger = DocumentActionName.DRAFT;
            }
        }
        DocumentStatusStateMachine stateMachine = documentStatusStateMachineBuilder.newStateMachine(currentState);
        stateMachine.fire(trigger);
        command.getStateSetter().accept(stateMachine.getCurrentState());
    }

}

File: Dddml.Wms.Common/Domain/InOut/InOutDocumentActionCommandHandler.cs
using Dddml.Wms.Specialization;
using Stateless;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Dddml.Wms.Domain.InOut
{
    public class InOutDocumentActionCommandHandler : IPropertyCommandHandler<string, string>
    {
        // //////////////////////////////////
        // 在异步处理模型下不能使用 ThreadStatic 来作为请求处理的上下文。
        //[ThreadStatic]
        //private static string _currentDocumentStatus;
        //private static StateMachine<string, string> _stateMachine;
        //static InOutDocumentActionCommandHandler()
        //{
        //    var tm = BuildStateMachine(() => _currentDocumentStatus, s => _currentDocumentStatus = s);
        //    _stateMachine = tm;
        //}

        private static StateMachine<string, string> BuildStateMachine(Func<String> stateAccessor, Action<String> stateMutator)
        {
            // //////////////////////////////////
            // TODO StateMachine 类库需要一个 StateMachineBuilder，避免重复配置。
            var tm = new StateMachine<string, string>(stateAccessor, stateMutator);

            tm.Configure(DocumentStatusIds.Initial)
                .Permit(DocumentAction.Draft, DocumentStatusIds.Drafted);

            tm.Configure(DocumentStatusIds.Drafted)
                .Permit(DocumentAction.Complete, DocumentStatusIds.Completed)
                .Permit(DocumentAction.Void, DocumentStatusIds.Voided);

            tm.Configure(DocumentStatusIds.Completed)
                .Permit(DocumentAction.Close, DocumentStatusIds.Closed)
                .Permit(DocumentAction.Reverse, DocumentStatusIds.Reversed);
            return tm;
        }

        public void Execute(IPropertyCommand<string, string> command)
        {
            var currentState = command.GetState();
            var trigger = command.Content;

            if (command.OuterCommandType == CommandType.Create)
            {
                if (String.IsNullOrWhiteSpace(currentState))
                { currentState = DocumentStatusIds.Initial; }
                if (trigger == null)
                { trigger = DocumentAction.Draft; }
            }

            var stateMachine = BuildStateMachine(() => currentState, command.SetState);
            stateMachine.Fire(trigger);
        }

    }
}
